chandra_aca/aca_image.py

```diff
@@ -309,6 +311,257 @@ def col0(self):
+            cls.flicker_cdf_x = np.linspace(hdr['cdf_x0'], hdr['cdf_x1'], hdr['n_cdf_x'])
+
+            # CDF bin range (e-/sec) for each for in flicker_cdfs.
+            cdf_bins = []
+            for ii in range(hdr['n_bin']):
+                cdf_bins.append(hdr[f'cdf_bin{ii}'])
+            cls.flicker_cdf_bins = np.array(cdf_bins)
+
+    def flicker_init(self, flicker_mean_time=10000, flicker_scale=1.0, seed=None):
+        """Initialize instance variables to allow for flickering pixel updates.
+
+        The ``flicker_scale`` can be interpreted as follows: if the pixel
+        was going to flicker by a multiplicative factor of (1 + x), now
+        make it flicker by (1 + x * flicker_scale).  This applies for flickers
+        that increase the amplitude.  For flickers that make the value smaller,
+        then it would be 1 / (1 + x) => 1 / (1 + x * flicker_scale).
+
+        The flicker_cdf file here was created using:
+        /proj/sot/ska/www/ASPECT/ipynb/chandra_aca/flickering-pixel-model.ipynb
+
+        Examples and performance details at:
+        /proj/sot/ska/www/ASPECT/ipynb/chandra_aca/flickering-implementation.ipynb
+
+        The model was reviewed and approved at SS&AWG on 2019-05-22.
+
+        :param flicker_mean_time: mean flickering time (sec, default=10000)
+        :param flicker_scale: multiplicative factor beyond model default for
+               flickering amplitude (default=1.0)
+        :param seed: random seed for reproducibility (default=None => no seed)
+        """
+        if not hasattr(self, 'flicker_cdf_bins'):
+            self._read_flicker_cdfs()
+
+        self.flicker_mean_time = flicker_mean_time
+        self.flicker_scale = flicker_scale
+        self.test_idx = 1 if seed == -1 else 0
+
+        if seed is not None and seed != -1:
+            np.random.seed(seed)
+            _numba_random_seed(seed)
+
+        # Make a flattened view of the image for easier update processing.
+        # Also store the initial pixel values, since flicker updates are
+        # relative to the initial value, not the current value (which would
+        # end up random-walking).
+        self.flicker_vals = self.view(np.ndarray).ravel()
+        self.flicker_vals0 = self.flicker_vals.copy()
+        self.flicker_n_vals = len(self.flicker_vals)
+
+        # Make a bool ACAImage like self to allow convenient mask/unmask of
+        # pixels to flicker.  This is used in annie.  Also make the corresponding
+        # 1-d ravelled version.
+        self.flicker_mask = ACAImage(np.ones(self.shape, dtype=np.bool),
+                                     row0=self.row0, col0=self.col0)
+        self.flicker_mask_vals = self.flicker_mask.view(np.ndarray).ravel()
+
+        # Get the index to the CDFs which is appropriate for each pixel
+        # based on its initial value.
+        self.flicker_cdf_idxs = np.searchsorted(self.flicker_cdf_bins,
+                                                self.flicker_vals0) - 1
```

Comment:
And I think this is the trickiest thing to check via code review; it is hard to say this indexing is right from this level, but if it passes whatever sanity checks you did while coding, I'm good with it.